extension/observer/dockerobserver/extension.go

```diff
@@ -85,173 +115,19 @@ func (d *dockerObserver) Shutdown(ctx context.Context) error {
-func (d *dockerObserver) updateEndpointsByContainerID(listener observer.Notify, cid string, latestEndpointsMap map[observer.EndpointID]observer.Endpoint) {
-	var removedEndpoints, addedEndpoints, updatedEndpoints []observer.Endpoint
-
-	if latestEndpointsMap != nil || len(latestEndpointsMap) != 0 {
-		// map of EndpointID to endpoint to help with lookups
-		existingEndpointsMap := make(map[observer.EndpointID]observer.Endpoint)
-		if endpoints, ok := d.existingEndpoints[cid]; ok {
-			for _, e := range endpoints {
-				existingEndpointsMap[e.ID] = e
-			}
-		}
-
-		// If the endpoint is present in existingEndpoints but is not
-		// present in latestEndpointsMap, then it needs to be removed.
-		for id, e := range existingEndpointsMap {
-			if _, ok := latestEndpointsMap[id]; !ok {
-				removedEndpoints = append(removedEndpoints, e)
-			}
+func (d *dockerObserver) ListEndpoints() []observer.Endpoint {
+	var endpoints []observer.Endpoint
+	for _, container := range d.dClient.Containers() {
+		for _, endpoint := range d.containerEndpoints(container.ContainerJSON) {
+			endpoints = append(endpoints, endpoint)
 		}
-
-		// if the endpoint is present in latestEndpointsMap, check if it exists
-		// already in existingEndpoints.
-		for _, e := range latestEndpointsMap {
-			// If it does not exist already, it is a new endpoint. Add it.
-			if existingEndpoint, ok := existingEndpointsMap[e.ID]; !ok {
-				addedEndpoints = append(addedEndpoints, e)
-			} else if !reflect.DeepEqual(existingEndpoint, e) {
-				// if it already exists, see if it's equal.
-				// if it's not equal, update it
-				// if its equal, no-op.
-				updatedEndpoints = append(updatedEndpoints, e)
-			}
-		}
-
-		// reset endpoints for this container
-		d.existingEndpoints[cid] = nil
-		// set the current known endpoints to the latest endpoints
-		for _, e := range latestEndpointsMap {
-			d.existingEndpoints[cid] = append(d.existingEndpoints[cid], e)
-		}
-	} else {
-		// if latestEndpointsMap is nil, we are removing all endpoints for the container
-		removedEndpoints = append(removedEndpoints, d.existingEndpoints[cid]...)
-		delete(d.existingEndpoints, cid)
-	}
-
-	if len(removedEndpoints) > 0 {
-		listener.OnRemove(removedEndpoints)
-	}
-
-	if len(addedEndpoints) > 0 {
-		listener.OnAdd(addedEndpoints)
-	}
-
-	if len(updatedEndpoints) > 0 {
-		listener.OnChange(updatedEndpoints)
 	}
+	return endpoints
 }
 
-// endpointsForContainer generates a list of observer.Endpoint given a Docker ContainerJSON.
+// containerEndpoints generates a list of observer.Endpoint given a Docker ContainerJSON.
 // This function will only generate endpoints if a container is in the Running state and not Paused.
-func (d *dockerObserver) endpointsForContainer(c *dtypes.ContainerJSON) map[observer.EndpointID]observer.Endpoint {
+func (d *dockerObserver) containerEndpoints(c *dtypes.ContainerJSON) map[observer.EndpointID]observer.Endpoint {
```

Comment:
Do we need to return a map here? It seems like it can `[]observer.Endpoint` based on the usage